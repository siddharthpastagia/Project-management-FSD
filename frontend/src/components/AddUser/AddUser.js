import React, { useState, useEffect } from "react";
import {
  Container,
  Row,
  Col,
  FormControl,
  Button,
  ListGroup,
  Alert
} from "react-bootstrap";
import { useFormik } from "formik";
import * as Yup from "yup";
import "./AddUser.scss";
import {
  getAllUsers,
  addNewUser,
  updateUserById,
  deleteUserById
} from "../../api/Api";

import * as _ from "lodash";

export const AddUser = () => {
  let [userUpdated, setUserUpdated] = useState(false);
  let [users, setUsers] = useState([]);
  let [userList, setUserList] = useState([]);
  let [sortMode, setSortMode] = useState(false);
  let [editMode, setEditMode] = useState(false);
  let [userId, setUserId] = useState("");
  let [firstName, setFirstName] = useState("");
  let [firstNameSort, setFirstNameSort] = useState(false);
  let [lastName, setLastName] = useState("");
  let [lastNameSort, setLastNameSort] = useState(false);
  let [empId, setEmpId] = useState("");
  let [empIdSort, setEmpIdSort] = useState(false);
  let [statusMessage, setStatusMessage] = useState({
    show: false,
    message: "",
    variant: ""
  });

  async function fetchAllUsers() {
    try {
      setUsers(await getAllUsers());
      setUserList(await getAllUsers());
      setUserUpdated(false);
    } catch (err) {
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: err,
        variant: "danger"
      });
    }
  }
  function editUser(user) {
    setEditMode(true);
    setUserId(user._id);
    setFirstName(user.firstName);
    setLastName(user.lastName);
    setEmpId(user.empId);
    setSortMode(false);
  }

  async function deleteUser(user) {
    try {
      const resp = await deleteUserById(user);
      setUserUpdated(true);
      setSortMode(false);
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: resp.message,
        variant: "success"
      });
    } catch (err) {
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: err,
        variant: "danger"
      });
    }
  }

  function resetFormState() {
    setFirstName("");
    setLastName("");
    setEmpId("");
    setEditMode(false);
  }

  useEffect(() => {
    fetchAllUsers();
  }, [userUpdated]);

  const handleChange = e => {
    let newList = [];
    if (e.target.value !== "") {
      newList = userList.filter(item => {
        const fData = item.firstName.toLowerCase();
        const lData = item.lastName.toLowerCase();
        const filter = e.target.value.toLowerCase();
        return (
          fData.includes(filter) ||
          lData.includes(filter) ||
          item.empId.includes(filter)
        );
      });
    } else {
      newList = userList;
    }
    setUsers(newList);
  };
  const handleSort = field => {
    setSortMode(true);
    const sortByField = _.sortBy(users, field);
    setUsers(sortByField);
  };
  const formik = useFormik({
    enableReinitialize: true,
    initialValues: {
      userId: userId,
      firstName: firstName,
      lastName: lastName,
      empId: empId
    },
    validationSchema: Yup.object({
      firstName: Yup.string().required("Please enter First Name"),
      lastName: Yup.string().required("Please enter Last Name"),
      empId: Yup.string().required("Please enter Employee ID")
    }),
    onSubmit: async value => {
      try {
        const resp = !editMode
          ? await addNewUser(value)
          : await updateUserById(value);
        setUserUpdated(true);
        setSortMode(false);
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: resp.message,
          variant: "success"
        });
      } catch (err) {
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: err,
          variant: "danger"
        });
      }
      resetFormState();
      formik.resetForm();
    }
  });
  return (
    <>
      <Container>
        <Row>
          <Col>
            <h3 className="mt-3">User Management</h3>
          </Col>
        </Row>
        <Row>
          <Col className="mt-2">
            <Alert
              variant={statusMessage.variant}
              show={statusMessage.show}
              onClose={() => {
                setStatusMessage({ ...statusMessage, show: false });
              }}
              dismissible
            >
              {statusMessage.message}
            </Alert>
            <form onSubmit={formik.handleSubmit}>
              <FormControl
                required
                placeholder="First Name"
                name="firstName"
                errors={formik.errors.firstName}
                className={
                  formik.touched.firstName
                    ? formik.errors.firstName
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
                {...formik.getFieldProps("firstName")}
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.firstName}
              </FormControl.Feedback>
              <FormControl
                placeholder="Last Name"
                name="lastName"
                required
                errors={formik.errors.lastName}
                {...formik.getFieldProps("lastName")}
                className={
                  formik.touched.lastName
                    ? formik.errors.lastName
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.lastName}
              </FormControl.Feedback>
              <FormControl
                placeholder="Employee ID"
                name="empId"
                required
                errors={formik.errors.empId}
                {...formik.getFieldProps("empId")}
                className={
                  formik.touched.empId
                    ? formik.errors.empId
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.empId}
              </FormControl.Feedback>
              <div className="text-center mt-4 mb-2">
                <Button
                  variant="primary"
                  disabled={!formik.isValid || !formik.dirty}
                  type="submit"
                  className="ml-2 mr-2"
                >
                  {editMode ? "Update" : "Add"}
                </Button>
                <Button
                  variant="outline-secondary"
                  onClick={() => {
                    resetFormState();
                    formik.resetForm();
                  }}
                >
                  Reset
                </Button>
              </div>
            </form>
          </Col>
        </Row>
        <hr />
        <Row>
          <Col xs={12} sm={6}>
            {users.length === 0 && (
              <Alert variant="warning">
                You don't have any users in the database!
              </Alert>
            )}
            <FormControl
              placeholder="Search"
              onChange={handleChange}
              className="mb-4"
            />
          </Col>
          <Col xs={12} sm={6}>
            Sort By:
            <Button
              variant="outline-primary"
              className={sortMode && firstNameSort ? "active ml-2" : "ml-2"}
              data-testid="sortFirstName"
              onClick={() => {
                handleSort("firstName");
                setFirstNameSort(true);
                setLastNameSort(false);
                setEmpIdSort(false);
              }}
            >
              First Name
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && lastNameSort ? "active ml-2" : "ml-2"}
              data-testid="sortLastName"
              onClick={() => {
                handleSort("lastName");
                setLastNameSort(true);
                setFirstNameSort(false);
                setEmpIdSort(false);
              }}
            >
              Last Name
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && empIdSort ? "active ml-2" : "ml-2"}
              data-testid="sortId"
              onClick={() => {
                handleSort("empId");
                setEmpIdSort(true);
                setFirstNameSort(false);
                setLastNameSort(false);
              }}
            >
              ID
            </Button>
          </Col>
        </Row>
        <Row className="mt-4">
          <Col>
            <ListGroup>
              {users.map(user => {
                return (
                  <ListGroup.Item className="user-list" key={user._id}>
                    <div>
                      <p>FirstName : {user.firstName}</p>
                      <p>LastName : {user.lastName}</p>
                      <p>Employee ID : {user.empId}</p>
                    </div>
                    <div>
                      <Button
                        variant="outline-primary"
                        onClick={() => editUser(user)}
                      >
                        Edit
                      </Button>
                      <br />
                      <Button
                        variant="outline-secondary"
                        className="mt-2"
                        onClick={() => {
                          deleteUser(user);
                        }}
                      >
                        Delete
                      </Button>
                    </div>
                  </ListGroup.Item>
                );
              })}
            </ListGroup>
          </Col>
        </Row>
      </Container>
    </>
  );
};
