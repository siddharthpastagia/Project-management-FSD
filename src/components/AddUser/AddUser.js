import React, { useState } from "react";
import {
  Container,
  Row,
  Col,
  FormControl,
  Button,
  ListGroup
} from "react-bootstrap";
import { useFormik } from "formik";
import * as Yup from "yup";
import "./AddUser.scss";

export const AddUser = () => {
  const usersList = [
    {
      firstName: "Sid",
      lastName: "pastagiya",
      empId: 393470
    },
    {
      firstName: "Palak",
      lastName: "pastagiya",
      empId: 393471
    },
    {
      firstName: "Zeeva",
      lastName: "pastagiya",
      empId: 393472
    }
  ];
  const formik = useFormik({
    initialValues: {
      firstName: "",
      lastName: "",
      empId: ""
    },
    validationSchema: Yup.object({
      firstName: Yup.string().required("Please enter First Name"),
      lastName: Yup.string().required("Please enter Last Name"),
      empId: Yup.string().required("Please enter Employee ID")
    })
  });
  const [users, setUsers] = useState(usersList);
  return (
    <>
      <Container>
        <Row>
          <Col className="mt-5">
            <form onSubmit={formik.handleSubmit}>
              <FormControl
                required
                placeholder="FirstName"
                name="firstName"
                errors={formik.errors.firstName}
                className={
                  formik.touched.firstName
                    ? formik.errors.firstName
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
                {...formik.getFieldProps("firstName")}
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.firstName}
              </FormControl.Feedback>
              <FormControl
                placeholder="LastName"
                name="lastName"
                required
                errors={formik.errors.lastName}
                {...formik.getFieldProps("lastName")}
                className={
                  formik.touched.lastName
                    ? formik.errors.lastName
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.lastName}
              </FormControl.Feedback>
              <FormControl
                placeholder="Employee ID"
                name="empId"
                required
                errors={formik.errors.empId}
                {...formik.getFieldProps("empId")}
                className={
                  formik.touched.empId
                    ? formik.errors.empId
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.empId}
              </FormControl.Feedback>
              <div className="float-right">
                <Button
                  variant="primary"
                  disabled={!formik.isValid || !formik.dirty}
                  type="submit"
                  className="ml-2 mr-2"
                >
                  Add
                </Button>
                <Button
                  variant="outline-secondary"
                  onClick={() => {
                    formik.resetForm();
                  }}
                >
                  Reset
                </Button>
              </div>
            </form>
          </Col>
        </Row>
        <hr />
        <Row>
          <Col>
            <ListGroup>
              {users.map(user => {
                return (
                  <ListGroup.Item className="user-list" key={user.empId}>
                    <div>
                      <p>FirstName : {user.firstName}</p>
                      <p>LastName : {user.lastName}</p>
                      <p>Employee ID : {user.empId}</p>
                    </div>
                    <div>
                      <Button variant="outline-primary">Edit</Button>
                      <br />
                      <Button
                        variant="outline-secondary"
                        className="mt-2"
                        onClick={() => {
                          const filterList = usersList.filter(
                            m => m.empId !== user.empId
                          );
                          setUsers(filterList);
                        }}
                      >
                        Delete
                      </Button>
                    </div>
                  </ListGroup.Item>
                );
              })}
            </ListGroup>
          </Col>
        </Row>
      </Container>
    </>
  );
};
