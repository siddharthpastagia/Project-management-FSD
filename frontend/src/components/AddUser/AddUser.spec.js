import React from "react";
import { MemoryRouter as Router } from "react-router-dom";
import { render, cleanup, wait, fireEvent } from "@testing-library/react";
import { toBeDisabled, toBeEmpty } from "@testing-library/jest-dom";
expect.extend({ toBeDisabled, toBeEmpty });

import { AddUser } from "./AddUser";
import axios from "axios";
jest.mock("axios");

describe("Add User Component", () => {
  const onSubmit = jest.fn();
  const fakeFirstName = "Siddharth";
  const fakeLastName = "Pastagiya";
  const fakeEmpId = "11";

  let userList = {
    data: [
      {
        _id: "5df53177f576f77b78e0e6d9",
        firstName: "User01",
        lastName: "01Last",
        empId: 1
      }
    ]
  };
  beforeEach(() => {
    axios.get.mockResolvedValue(userList);
  });
  afterEach(cleanup);
  it("Should render Add User component", async () => {
    const { container } = render(
      <Router>
        <AddUser />
      </Router>
    );
    await wait(() => {
      expect(container).toMatchSnapshot();
    });
  });
  it("Get user List ", async () => {
    const { getByText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    expect(getByText("You don't have any users in the database!"));
  });

  it("Should able to Add User ", async () => {
    const { getByText, getByPlaceholderText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    const firstName = getByPlaceholderText("First Name");
    const lastName = getByPlaceholderText("Last Name");
    const empId = getByPlaceholderText("Employee ID");
    const submitBtn = getByText("Add");

    fireEvent.change(firstName, {
      target: { name: "firstName", value: fakeFirstName }
    });
    fireEvent.change(lastName, {
      target: { name: "lastName", value: fakeLastName }
    });
    fireEvent.change(empId, { target: { name: "empId", value: fakeEmpId } });

    await wait(() => {
      expect(submitBtn).not.toBeDisabled();
      fireEvent.click(submitBtn);
    });
  });

  it("Should able to Edit User ", async () => {
    const { getByText, getByPlaceholderText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    await wait(() => {
      const firstName = getByPlaceholderText("First Name");
      const editBtn = getByText("Edit");

      fireEvent.click(editBtn);

      const updateBtn = getByText("Update");
      fireEvent.change(firstName, {
        target: { name: "firstName", value: fakeFirstName }
      });

      expect(updateBtn).not.toBeDisabled();
      fireEvent.click(updateBtn);
    });
  });

  it("Should able to Delete User ", async () => {
    const { getByText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    await wait(() => {
      const DeleteBtn = getByText("Delete");
      expect(DeleteBtn).not.toBeDisabled();
      fireEvent.click(DeleteBtn);
    });
  });

  it("Should able to Reset User Form ", async () => {
    const { getByText, getByPlaceholderText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    const firstName = getByPlaceholderText("First Name");
    fireEvent.change(firstName, {
      target: { name: "firstName", value: fakeFirstName }
    });
    const resetBtn = getByText("Reset");
    expect(resetBtn).not.toBeDisabled();
    fireEvent.click(resetBtn);
    expect(firstName).toBeEmpty();
  });

  it("Should able to Serach the user from user List ", async () => {
    const { getByText, getByPlaceholderText } = render(
      <Router>
        <AddUser />
      </Router>
    );
    await wait(() => {
      const searchInp = getByPlaceholderText("Search");
      fireEvent.change(searchInp, {
        target: { name: "searchUser", value: "User" }
      });
    });
    expect(getByText("FirstName : User01"));
  });
  it("Should able to click on sorting options ", async () => {
    const { getByText, getByTestId } = render(
      <Router>
        <AddUser />
      </Router>
    );
    await wait(() => {
      const sortFirstNameBtn = getByTestId("sortFirstName");
      const sortLastNameBtn = getByTestId("sortLastName");
      const sortIdBtn = getByTestId("sortId");

      fireEvent.click(sortFirstNameBtn);
      fireEvent.click(sortLastNameBtn);
      fireEvent.click(sortIdBtn);
      expect(sortFirstNameBtn).not.toBeDisabled();
    });
    expect(getByText("FirstName : User01"));
  });
});
