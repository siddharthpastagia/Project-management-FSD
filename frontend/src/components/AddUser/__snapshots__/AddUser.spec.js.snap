// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add User Component Should render Add User component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h3
          class="mt-3"
        >
          User Management
        </h3>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="mt-2 col"
      >
        <form>
          <input
            class="mb-2 mt-2 form-control"
            name="firstName"
            placeholder="First Name"
            required=""
            value=""
          />
          <div
            class="invalid-feedback"
          />
          <input
            class="mb-2 mt-2 form-control"
            name="lastName"
            placeholder="Last Name"
            required=""
            value=""
          />
          <div
            class="invalid-feedback"
          />
          <input
            class="mb-2 mt-2 form-control"
            name="empId"
            placeholder="Employee ID"
            required=""
            value=""
          />
          <div
            class="invalid-feedback"
          />
          <div
            class="text-center mt-4 mb-2"
          >
            <button
              class="ml-2 mr-2 btn btn-primary"
              disabled=""
              type="submit"
            >
              Add
            </button>
            <button
              class="btn btn-outline-secondary"
              type="button"
            >
              Reset
            </button>
          </div>
        </form>
      </div>
    </div>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-sm-6 col-12"
      >
        <input
          class="mb-4 form-control"
          placeholder="Search"
        />
      </div>
      <div
        class="col-sm-6 col-12"
      >
        Sort By:
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortFirstName"
          type="button"
        >
          First Name
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortLastName"
          type="button"
        >
          Last Name
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortId"
          type="button"
        >
          ID
        </button>
      </div>
    </div>
    <div
      class="mt-4 row"
    >
      <div
        class="col"
      >
        <div
          class="list-group"
        >
          <div
            class="user-list list-group-item"
          >
            <div>
              <p>
                FirstName : 
                User01
              </p>
              <p>
                LastName : 
                01Last
              </p>
              <p>
                Employee ID : 
                1
              </p>
            </div>
            <div>
              <button
                class="btn btn-outline-primary"
                type="button"
              >
                Edit
              </button>
              <br />
              <button
                class="mt-2 btn btn-outline-secondary"
                type="button"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
